### DISCUSSION TITLE
Granting CRUD permissions to a user to manage auth.users

### DISCUSSION THREAD

**Discussion by mburakkalkan** [supabase/supabase#26973] [Questions] [answered]
Hello, I want to implement user management in my app but couldn't find a reliable solution for that.

As I understand, only the service_role token can modify the auth.users table. This is ok for single admin solutions, but I want to grant crud rights to some of my users for auth.users table without giving them the service_role key.

Is that possible and how can I implement that?

**Comment by GaryAustin1** [comment]
Service_role can only modify things thru the user auth.admin calls.  There is no access to the auth schema from the API for any role directly.
What type of user operations are you planning to do?  Many times this can be done on other tables like a profile table or a roles table that is linked to auth.users, but can be accessed from the API.

**Reply by mburakkalkan** [reply]
Thanks @GaryAustin1, yes I already have profiles and roles management over a profiles table, but that's not enough for my requirements. 

I should be able to grant permissions for the users who has the "admin" role in my app, so that they can edit, delete or create new users. But I couldn't find a way to do that with the current supabase scenario.

**Reply by GaryAustin1** [reply]
Only service role jwt can call auth.admin user calls.  If you are doing the manipulation serverside you would have a role admin for some users with either a role table or something like auth hookshttps://supabase.com/docs/guides/database/postgres/custom-claims-and-role-based-access-control-rbac to allow certain users to run the server code doing auth.admin calls. 

Manipulating the auth.users table directly is not recommended and would involve having to create grants for a postgres user role to do that.  And it can't be done from the REST API directly, although you could use rpc calls to security definer functions.

**Reply by mburakkalkan** [reply]
Got it @GaryAustin1, but what I can't understand is that is a very common scenario and I think the implementation shouldn't be complex like that.

Consider creating an app for a small company, and the company have users with different roles like director, assistant directors, supervisors and workers. And all of these people are the "users" for the app, right? So, in this case, directors or assistant directors (or someone else...) should be able to manipulate these users table. Because as a developer, you are just responsible for creating or developing the app, not managing the business.

As a result, since it is obvious that I cannot take the risk as giving the `service_role` key to these non-technical people, can we say that "supabase is just for end user apps and not for corporate or enterprise applications"?

**Reply by GaryAustin1** [reply, marked as answer]
You would not give users the service role.   You would have serverside, or edge function code which checks if the user is a certain role from a table or auth hooks/custom_claims in the jwt.  If the user has an admin role then you would do the auth.admin calls you need to carry out the server operation.   

You can create other roles and do database operations from database security definer functions also, but it is still recommended to do auth.users operations directly.  What you cannot do is access auth schema from the API directly.

**Reply by mburakkalkan** [reply]
I tried some approaches over security definer functions but I couldn't get success. And I was searching a solution which won't need any serverside resource but yes, edge functions may help at this point.

I will try, thanks for your help @GaryAustin1
